Gemfile

```diff
@@ -6,6 +6,8 @@ git_source(:github) do |repo_name|
 end
 
 gem "rails", "~> 5.1.4"
+gem "config"
+gem "bcrypt", "3.1.11"
```

Comment:
gem này để làm gì vậy em? tại sao phải có nó?

Reply:
bcrypt là môt thuật toán mã hóa một chiều, gem "bcrypt" là dùng để mã hóa mật khẩu người dùng và kiểm tra mật khẩu khi đăng nhâp.

Reply:
tại sao cần phải khai báo gem này?

Reply:
has_secure_password được sử dụng trong user.rb là một method, và method này yêu cầu bcrypt, nên ta cần khai báo gem bcrypt này để sử dụng. Ví dụ cụ thể method has_secure_password: 


def has_secure_password
  gem 'bcrypt-ruby', '~> 3.0.0'
  require 'bcrypt'

  attr_reader :password

  validates_confirmation_of :password
  validates_presence_of     :password_digest

  include InstanceMethodsOnActivation

  if respond_to?(:attributes_protected_by_default)
    def self.attributes_protected_by_default
      super + ['password_digest']
    end
  end
end